File: backend/routes/requirements.py
from fastapi import APIRouter, Depends, HTTPException
from repositories.requirements import requirement_repo
from schemas.requirement import RequirementCreate, RequirementUpdate, RequirementResponse
from utils.response import ApiResponse

router = APIRouter(prefix="/requirements", tags=["Requirements"])


@router.get("", response_model=ApiResponse)
async def get_requirements(project_id: int = None):
    """Get all requirements, optionally filtered by project"""
    if project_id:
        requirements = requirement_repo.get_by_project(project_id)
    else:
        requirements = requirement_repo.get_all()
    return ApiResponse.ok(data=requirements)


@router.get("/{requirement_id}", response_model=ApiResponse)
async def get_requirement(requirement_id: int):
    """Get a requirement by ID"""
    requirement = requirement_repo.get_by_id(requirement_id)
    if not requirement:
        raise HTTPException(status_code=404, detail="Requirement not found")
    return ApiResponse.ok(data=requirement)


@router.post("", response_model=ApiResponse)
async def create_requirement(requirement_in: RequirementCreate):
    """Create a new requirement"""
    requirement = requirement_repo.create(requirement_in.model_dump())
    return ApiResponse.ok(
        data=requirement,
        message="Requirement created",
    )


@router.put("/{requirement_id}", response_model=ApiResponse)
async def update_requirement(
    requirement_id: int,
    requirement_in: RequirementUpdate,
):
    """Update a requirement"""
    requirement = requirement_repo.get_by_id(requirement_id)
    if not requirement:
        raise HTTPException(status_code=404, detail="Requirement not found")
    updated = requirement_repo.update(
        requirement_id, requirement_in.model_dump(exclude_unset=True)
    )
    return ApiResponse.ok(
        data=updated, message="Requirement updated"
    )


@router.delete("/{requirement_id}", response_model=ApiResponse)
async def delete_requirement(requirement_id: int):
    """Delete a requirement"""
    requirement = requirement_repo.get_by_id(requirement_id)
    if not requirement:
        raise HTTPException(status_code=404, detail="Requirement not found")
    requirement_repo.delete(requirement_id)
    return ApiResponse.ok(message="Requirement deleted")

File: backend-nodejs/src/routes/requirements.js
/**
 * Requirements Routes
 */
const { BaseRepository } = require('../repositories/base');

const repo = new BaseRepository('requirements.json');

/**
 * Format success response
 */
function successResponse(data = null, message = 'Success') {
  return { success: true, message, data };
}

/**
 * Format error response
 */
function errorResponse(message, error = null) {
  return { success: false, message, error, data: null };
}

/**
 * Register requirements routes
 */
async function requirementsRoutes(fastify) {
  // Get all requirements
  fastify.get('/', async (request, reply) => {
    try {
      const requirements = await repo.findAll();
      return successResponse(requirements);
    } catch (error) {
      request.log.error(error);
      return errorResponse('Failed to get requirements');
    }
  });

  // Get requirement by ID
  fastify.get('/:id', async (request, reply) => {
    try {
      const requirementId = parseInt(request.params.id, 10);
      const requirement = await repo.findById(requirementId);
      if (!requirement) {
        reply.code(404);
        return errorResponse('Requirement not found');
      }
      return successResponse(requirement);
    } catch (error) {
      request.log.error(error);
      reply.code(500);
      return errorResponse('Failed to get requirement');
    }
  });

  // Create requirement
  fastify.post('/', async (request, reply) => {
    try {
      const requirement = await repo.create(request.body);
      return successResponse(requirement, 'Requirement created');
    } catch (error) {
      request.log.error(error);
      reply.code(500);
      return errorResponse('Failed to create requirement');
    }
  });

  // Update requirement
  fastify.put('/:id', async (request, reply) => {
    try {
      const requirementId = parseInt(request.params.id, 10);
      const updated = await repo.update(requirementId, request.body);
      if (!updated) {
        reply.code(404);
        return errorResponse('Requirement not found');
      }
      return successResponse(updated, 'Requirement updated');
    } catch (error) {
      request.log.error(error);
      reply.code(500);
      return errorResponse('Failed to update requirement');
    }
  });

  // Delete requirement
  fastify.delete('/:id', async (request, reply) => {
    try {
      const requirementId = parseInt(request.params.id, 10);
      const deleted = await repo.delete(requirementId);
      if (!deleted) {
        reply.code(404);
        return errorResponse('Requirement not found');
      }
      return successResponse(null, 'Requirement deleted');
    } catch (error) {
      request.log.error(error);
      reply.code(500);
      return errorResponse('Failed to delete requirement');
    }
  });
}

module.exports = requirementsRoutes;
